DS-3883: Speed up Item summary lists by being smarter about when we load Thumbnails/Bitstreams

tdonohue opened the pull request:
Fixes https://jira.duraspace.org/browse/DS-3883

The Collection/Community performance issues described in this ticket are caused by Items that have a large number of Thumbnails (e.g. Items with many PDF bitstreams or Image bitstreams, where each bitstream has a corresponding Thumbnail).

As a quick fix for this scenario, this PR improves the logic used in the XMLUI `ItemAdapter`. This is the class that generates the Item specific mets.xml files (e.g. http://demo.dspace.org/xmlui/metadata/handle/10673/5/mets.xml) that are used to gather metadata/bitstream information to be used by XMLUI themes.

On Collection / Community homepages (or anywhere an Item Summary is used), the XMLUI accesses this mets.xml document using a special querystring, so that *only* metadata and thumbnails are returned, e.g. http://demo.dspace.org/xmlui/metadata/handle/10673/5/mets.xml?sections=dmdSec,fileSec&fileGrpTypes=THUMBNAIL

Therefore, this PR updates the `ItemAdapter` logic as follows:
* If the `fileGrpTypes` querystring param includes `THUMBNAIL` but NOT the `ORIGINAL` bundle, then _only include the main thumbnail in the mets.xml_.  This ensures that on Item Summary listings only *one* thumbnail is ever loaded.
* In the above scenario, also *avoid* linking the thumbnail bitstream back to its ORIGINAL bitstream in the mets.xml (as that linkage serves no purpose, since the ORIGINAL bitstream isn't even in the mets.xml). This ensures that we do not loop through all the ORIGINAL bitstreams unless they are explicitly included in the list of `fileGrpTypes`

With these two fixes in place, the performance of the Community/Collection homepages is back to normal (and may even be *faster* than in v5.x, as we are only ever loading a single Thumbnail per Item).

(SIDENOTE: it is worth noting that this PR _only improves performance for Item Summary lists_.  On the Item Homepage, performance is still going to be extremely slow for Items having lots of Bitstreams & Thumbnails. This is because the Item homepage always lists all Bitstreams/Thumbnails. We could continue to investigate ways to improve this performance via fewer database queries, better caching, etc. But, it's worth noting this also may be fixable via some form of pagination of Bitstreams -- which is planned for the 7.x REST API . Small performance improvements may also be achieved by customizing the XMLUI Theme to only display a limited set of Bitstreams on the Item page. )

tdonohue committed:
DS-3883: Don't loop through original bitstreams if only displaying thumbnails

tdonohue committed:
DS-3883: If only including thumbnails, only load the main item thumbnail.

terrywbrady commented on dspace-xmlui/src/main/java/org/dspace/app/xmlui/objectmanager/ItemAdapter.java:
```diff
@@ -761,20 +764,41 @@ protected void renderFileSection(Context context) throws SQLException, SAXExcept
+            // If this is the THUMBNAIL bundle, and we are NOT including content bundle,
+            // Then assume this is an item summary page, and we can just include the main thumbnail.
+            if (isDerivedBundle && !includeContentBundle)
+            {
```
The derived bundle could be TEXT or THUMBNAIL.  Should this check explicitly for the thumbnail bundle?

kshepherd replied:
+1 . mind you if you look a few lines down, if it's TEXT it'll return null on the ItemService.getThumbnail() call and so the bitstream won't be added to the bitstreams list anyway

kshepherd replied:
oh, no, i'm wrong about that.. it might mean the thumbnail is added twice if the item has both THUMBNAIL and TEXT since the 'getThumbnail' is really just against the item and not the bundle..

terrywbrady reviewed (approved):
I added one question in the code review.

The code looks good and I have verified the performance improvement on my server.

This seems like an appropriate containment of the fix for the 6.3 release.  I agree that further refinement of the hibernate query/cache process makes sense going forward.

kshepherd reviewed (approved):
+1, thanks @tdonohue , I've tested this successfully and my vagrant vm returns collection home with items containing >600 bitstreams in under a second. Glad we got to the bottom of this one! i agree with @terrywbrady 's comment that being slightly more narrow with the 'isDerived' test to exclude TEXT and other derived bundles, will speed things up even further

tdonohue commented:
@terrywbrady and @kshepherd : Thanks for the feedback!  I corrected that one line of code that you commented on (and squashed that change into my second commit).

alanorth commented:
Wow, this is a great win. As someone stuck on DSpace 5.x with lots of thumbnails I'd love to have this back-ported there too.
